src/semver.toit

```diff
@@ -1,195 +1,799 @@
+
+    if not a-prerelease and not b-prerelease:
+      return if-equal.call
+
+    // Any prerelease is lower than no prerelease.
+    if not a-prerelease: return 1
+    if not b-prerelease: return -1
+
+    comp := compare-dotted_ a-prerelease b-prerelease
+    if comp != 0: return comp
+    return if-equal.call
+  */
+
+  is-letter_ c/int -> bool:
+    return 'A' <= c <= 'Z' or 'a' <= c <= 'z'
+
+  is-digit_ c/int -> bool:
+    return '0' <= c <= '9'
+
+  is-non-digit_ c/int -> bool:
+    return c == '-' or is-letter_ c
+
+  is-identifier-character_ c/int -> bool:
+    return is-digit_ c or is-non-digit_ c
+
+  is-valid-build-metadata_ build/string -> bool:
+    parts := build.split "."
+    if parts.size == 0: return false
+    parts.do: | part/string |
+      if part.is-empty: return false
+      part.do:
+        if not is-identifier-character_ it: return false
+    return true
+
+  is-valid-prerelease_ prerelease/string -> bool:
+    parts := prerelease.split "."
+    if parts.size == 0: return false
+    parts.do: | part/string |
+      if part.is-empty: return false
+      only-digits := true
+      // Either an alpha-num identifier, or a numeric identifier.
+      // Numeric identifiers must not have leading zeros.
+      part.do:
+        if not is-digit_ it: only-digits = false
+        if not is-identifier-character_ it: return false
+      if only-digits and part.size > 1 and part[0] == '0': return false
+    return true
+
+  /**
+  Returns true if $str is a valid semver string.
+
+  If $allow-v is true, then the string may start with a 'v' or 'V'.
+  If $require-major-minor-patch is true, then the string must have at least a
+    major, minor and patch version. Otherwise it is enough to have a major
+    version (or a major and minor version).
+  */
+  /* propose deletion
+  is-valid str/string --allow-v/bool=true --require-major-minor-patch/bool=true -> bool:
+    if allow-v and (str.starts-with "v" or str.starts-with "V"):
+      str = str[1..]
+
+    build-index := str.index-of "+"
+    if build-index != -1:
+      build := str[build-index + 1..]
+      if not is-valid-build-metadata_ build: return false
+      str = str[..build-index]
+
+    prerelease-index := str.index-of "-"
+    if prerelease-index != -1:
+      prerelease := str[prerelease-index + 1..]
+      if not is-valid-prerelease_ prerelease: return false
+      str = str[..prerelease-index]
+
+    version-core := str
+
+    parts := version-core.split "."
+
+    if parts.size == 0: return false
+    if parts.size > 3: return false
+    if require-major-minor-patch and parts.size != 3: return false
+
+    parts.do: | part/string |
+      if part.is-empty: return false
+      part.do:
+        if not is-digit_ it: return false
+      if part.size > 1 and part[0] == '0': return false
+
+    return true
+  */
+
+// ORIGINAL PEG PARSER CODE BELOW HERE -----------------------------------------
+
+class SemanticVersionPEGParser extends parser.PegParserBase_:
```

Comment:
Remove from this file.
We should not use any private classes from the core libraries. if we change the name (or remove it), then this package would stop working.
If you really want to keep it for testing, then move it to the tests directory. There, it wouldn't hurt users if it broke.